web/src/stream_data.ts

```diff
@@ -477,6 +477,59 @@ export function is_new_stream_announcements_stream_muted(): boolean {
     return is_muted(realm.realm_new_stream_announcements_stream_id);
 }
 
+// This function will be true for every case since the server should be
+// preventing a StreamSubscription from reaching clients without
+// metadata access.
+// This function can be used to allow callers to log blueslip errors
+// when the client seems to have a group it shouldn't have access to,
+// in order to find server bugs.
+export function has_metadata_access(sub: StreamSubscription): boolean {
+    if (sub.is_web_public) {
+        return true;
+    }
+
+    if (page_params.is_spectator) {
+        return false;
+    }
+
+    if (!current_user.is_guest && !sub.invite_only) {
+        return true;
+    }
+
+    if (sub.subscribed) {
+        return true;
+    }
+
+    if (can_administer_accessible_channel(sub)) {
+        return true;
+    }
+
+    // Users that can add other subscribers to a private channel
+    // have content access to that channel. Having content access
+    // should give them metadata access to that private channel even
+    // when unsubscribed.
+    const can_add_subscribers = user_groups.is_user_in_setting_group(
+        sub.can_add_subscribers_group,
+        people.my_current_user_id(),
+    );
+    if (can_add_subscribers) {
+        return true;
+    }
+
+    return false;
+}
+
+function can_administer_accessible_channel(sub: StreamSubscription): boolean {
```

Comment:
I am somewhat confused by this name. A channel is accessible, i.e. metadata access, if user is part of `can_administer_channel_group`, so wouldn't `can_administer_channel` be a good name?

Reply:
Yeah, I think this function name probably should be replaced -- it dates from before the "content access"/"metadata access" naming scheme. `can_administer_channel` is probably reasonable for the web app use case?

Reply:
Did this, plus of course the hunks for callers to use the new name.

``` diff
     return true;
 }
 
-function can_administer_accessible_channel(sub: StreamSubscription): boolean {
+function can_administer_channel(sub: StreamSubscription): boolean {
+    // Note that most callers should use wrappers like
+    // can_change_permissions_requiring_content_access, since actions
+    // that can grant access to message content require content access
+    // in addition to being a channel administrator.
     if (current_user.is_admin) {
         return true;
     }
```